framework/include/timesteppers/Stepper.h

```diff
@@ -0,0 +1,606 @@
+    }
+  }
+  double val(double dt)
+  {
+    if (on)
+      DBG << "[LOG] " << std::string(level*4, ' ') << "* dt=" << dt << "\n";
+    return dt;
+  }
+  static int level;
+  static bool on;
+  std::string _loc;
+};
+int Logger::level = 0;
+bool Logger::on = true;
+
+// IterationAdaptiveDT stepper can be approximated something like this:
+//
+//     new MinOf(
+//         new FixedPointStepper(sync_times, time_tol),
+//         new MinOf(
+//             new FixedPointStepper(time_list, time_tol),
+//             new ConstrFuncStepper(
+//                 new DTLimitStepper(
+//                     RetryUnusedStepper(
+//                         AlternatingStepper(
+//                             new PiecewiseStepper(time_list, dt_list),
+//                             new AdaptiveStepper([config...]),
+//                             time_list,
+//                             time_tol
+//                         )
+//                     ),
+//                     dt_min,
+//                     dt_max
+//                 ),
+//                 constr_func,
+//                 max_diff
+//             )
+//         )
+//     );
+
+// Original AB2PredictorCorrector stepper behavior can be achieved by wrapping
+// steppers roughly like so:
+//
+//     new EveryNStepper(
+//         new MaxRatioStepper(new PredictorCorrector(...), max_ratio),
+//         every_n
+//     )
+
+struct StepperInfo {
+  // The number of times the simulation has requested a dt.
+  // This is equal to one on the first call fo advance(...).
+  int step_count;
+
+  // time info
+  double time;
+  double prev_dt;
+  double prev_prev_dt;
+  std::string time_integrator;
+
+  // solve stats
+  unsigned int nonlin_iters;
+  unsigned int lin_iters;
+  bool converged;
+  bool prev_converged;
+  double solve_time_secs;
+
+  // solution stuff
+  NumericVector<Number>* soln_nonlin;
+  NumericVector<Number>* soln_aux;
+  NumericVector<Number>* soln_predicted;
+
+  // backup/restore
+  bool sched_backup;
+  bool sched_restore;
+  double restore_time;
+};
+
+class Stepper
+{
+public:
+  // Implementations of advance should strive to be idempotent.  Return the
+  // next value to use as dt.
+  virtual double advance(const StepperInfo* si) = 0;
+};
+
+/////////////////////////////////////////////////////////////
+///////////////// Stepper Implementations ///////////////////
+/////////////////////////////////////////////////////////////
+
+// Always returns a single, unchanging, user-specified dt.
+class ConstStepper : public Stepper
+{
+public:
+  ConstStepper(double dt) : _dt(dt)
+  {
+  }
+
+  virtual double advance(const StepperInfo* si)
+  {
+    return _dt;
+  }
+
+private:
+  double _dt;
+};
+
+// Calls an underlying stepper to compute dt. Modifies dt to be within
+// specified fixed lower and upper bounds (or throws an error - depending on
+// configuration).
+class DTLimitStepper : public Stepper
```

Comment:
Also: try to keep stuff out of headers... already takes forever to compile MOOSE